frontend/src/app/offers/create-offer/create-offer.component.ts

```diff
@@ -0,0 +1,154 @@
+import { ActivatedRoute, Params } from '@angular/router';
+import { environment } from 'environments/environment';
+import { Subscription } from 'rxjs/Subscription';
+import { FileReaderEvent, FileReaderEventTarget } from '../../models';
+import { NgForm, FormGroup } from '@angular/forms';
+import { FormBuilder } from '@angular/forms';
+import { Validators } from '@angular/forms';
+import { Image } from 'app/homepage-offer/image.model';
+
+@Component({
+  selector: 'volontulo-create-offer',
+  templateUrl: './create-offer.component.html',
+  styleUrls: ['./create-offer.component.scss']
+})
+export class CreateOfferComponent implements OnInit, OnDestroy {
+  public djangoRoot = environment.djangoRoot;
+  public file: File;
+  public hasOrganization = false;
+  public inEditMode = false;
+  public isAdmin = false;
+  public offer: AppOffer = new AppOffer;
+  public reader = new FileReader();
+  public user: User;
+  public userSubscription: Subscription;
+  public form: FormGroup;
+
+  constructor(
+    private authService: AuthService,
+    private offersService: OffersService,
+    private route: ActivatedRoute,
+    private fb: FormBuilder,
+  ) {}
+
+  ngOnInit() {
+    this.form = this.fb.group({
+      title: ['', Validators.required],
+      location: ['', Validators.required],
+      organization: ['', Validators.required],
+      startedAt: ['', Validators.required],
+      actionOngoing: ['', Validators.required],
+      finishedAt: ['', Validators.required],
+      constantCoop: [],
+      recruitmentStartDate: ['', Validators.required],
+      recruitmentEndDate: ['', Validators.required],
+      volunteersLimit: [],
+      reserveRecruitmentStartDate: [],
+      reserveRecruitmentEndDate: [],
+      reserveVolunteersLimit: [],
+      description: ['', Validators.required],
+      timeCommitment: ['', Validators.required],
+      benefits: ['', Validators.required],
+      requirements: ['', Validators.required],
+      image: [],
+    })
+
+      this.userSubscription = this.authService.user$
+      .subscribe(
+        response => {
+          this.user = response;
+            this.isAdmin = this.user['isAdministrator'];
+            this.hasOrganization = this.user['organizations'].length > 0;
+          }
+      );
+
+    this.route.params
+    .map(params => params.offerId)
+    .filter(offerId => offerId !== undefined)
+    .switchMap(offerId => this.offersService.getOffer(offerId))
+    .do(offer => this.toDataUrl(offer.image, (base64) => this.offer.image = {
+      content: base64,
+      filename: 'image.jpg'
+    }))
+    .subscribe((response: BaseOffer) => {
+      this.offer = response as AppOffer;
+      this.inEditMode = true;
+      this.form.patchValue(this.offer);
+    });
+  }
+
+  ngOnDestroy() {
+    if (this.userSubscription) {
+      this.userSubscription.unsubscribe();
+    }
+  }
+
+  toDataUrl(url, callback) {
+    const xhr = new XMLHttpRequest();
```

Comment:
Could we rewrite it to HttpClient like in services - `this.http.get(url, {responseType: 'blob'}).subscrible( // image processing here)` or something similar? I know it is probably possible, you can check Angular docs how to pass the 'blob' type in configuration object.

HttpClient is actually Angular's wrapper over old-school XMLHttpRequest and it internally uses it anyway.

Reply:
I hope you didn't forget about the comment above :) You can import http module like this
`import { HttpClient } from '@angular/common/http';` . We switched for the new HttpClient recently, but it works pretty much the same as the old one.